### PR TITLE
Adding redux

### PR DESCRIPTION
**Pull request by Ofir-Smuha**
## Adding redux
- projects / timeTracking state moved into redux

### CODE REVIEW

File: proj/src/actions/projectsActions.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import {OPEN_EDIT_PROJECT, DELETE_PROJECT, CLOSE_EDIT_PROJECT, IS_LOADING, EDIT_PROJECT, ADD_PROJECT} from 'actions/types'
```
Comment (ronlanger) [inline proj/src/actions/projectsActions.js]:
```suggestion
import * as AT from 'actions/types'
```

File: proj/src/actions/projectsActions.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+export const addProject = (project) => ({
+  type: ADD_PROJECT,
```
Comment (ronlanger) [inline proj/src/actions/projectsActions.js]: and then 
```suggestion
  type: AT.ADD_PROJECT,
```

File: proj/src/components/projects/EditProject.js [hunk 3]
```diff
@@ -1,5 +1,10 @@
-import React from 'react'
-import styled from 'styled-components'
+import React from 'react';
+import styled from 'styled-components';
+import { connect } from 'react-redux';
+import uuidv4 from 'uuid/v4';
+import PropTypes from 'prop-types';
+
+import { closeEditModal, isLoading, addProject, editProject} from 'actions/projectsActions';
```
Comment (ronlanger) [inline proj/src/components/projects/EditProject.js]: you can use import * as projectActions here if you have to import many actions. it also prevent the confusion of using the imported action creator or dispatching the action from props

File: proj/src/components/projects/EditProject.js [hunk 4]
```diff
@@ -59,25 +64,62 @@ const CloseButton = styled.button`
...
+      const newProjectId = uuidv4()
+      const newProject = {id: newProjectId, name: inputValue}
+      props.addProject(newProject)
+    }
+  }
+
+  const renderTitle = () => {
+    if (props.currProject) {
+      return <Title>{props.currProject.name}</Title>
+    } else {
```
Comment (ronlanger) [inline proj/src/components/projects/EditProject.js]: no need for else here

File: proj/src/components/projects/EditProject.js [hunk 5]
```diff
@@ -59,25 +64,62 @@ const CloseButton = styled.button`
...
+      return <Title>Add new project</Title>
+    }
+  }
+
+  const renderInput = () => {
+    if (props.currProject) {
+      return <Input type="text" name="name" defaultValue={props.currProject.name}/>
+    } else {
```
Comment (ronlanger) [inline proj/src/components/projects/EditProject.js]: no need for else here

File: proj/src/components/projects/EditProject.js [hunk 6]
```diff
@@ -59,25 +64,62 @@ const CloseButton = styled.button`
...
+  closeEditModal: PropTypes.func
+}
+
+const mapStateToProps = state => ({
+  currProject: state.projects.currProject
+})
 
-export default EditProject
\ No newline at end of file
+export default connect(mapStateToProps, { closeEditModal, isLoading, addProject, editProject })(EditProject)
```
Comment (ronlanger) [inline proj/src/components/projects/EditProject.js]: if you imported * as projectActions you would need to change here as well
```suggestion
export default connect(mapStateToProps, { 
    closeEditModal: projectActions.closeEditModal, 
    .... // complete the rest here
})(EditProject)
```

again, this is just a suggestion. another advantage of this besides name conflicts that come to mind is that you know from where each action creator came from. the downside is that it is more to write

File: proj/src/reducers/projects/projectsReducer.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    projects: {
+        1: {id: 1, name: 'Project 1'},
+        2: {id: 2, name: 'Project 2'},
+        3: {id: 3, name: 'Project 3'},
+    }
+};
+
+
+export default handleActions({
+    [combineActions(ADD_PROJECT, EDIT_PROJECT,)]: (state, payload) =>
```
Comment (ronlanger) [inline proj/src/reducers/projects/projectsReducer.js]: would be better to put combined actions vertically so it's more readable
```suggestion
    [combineActions(
        ADD_PROJECT, 
        EDIT_PROJECT
    )]: (state, payload) =>
```

File: proj/src/reducers/root.js [hunk 8]
```diff
@@ -0,0 +1,13 @@
+import { combineReducers } from 'redux';
+
+import loginReducer from 'reducers/login/loginReducer'
+import projectsReducer from 'reducers/projects/projectsReducer'
+import statistics from 'reducers/statistics/statisticsReducer'
+import timeTrackingReducer from 'reducers/time-tracking/timeTrackingReducer'
+
+export default combineReducers({
```
Comment (ronlanger) [inline proj/src/reducers/root.js:8]: we need to talk about naming the reducers so we will avoid things like state.projects.projects in mapStateToProps

File: proj/src/reducers/time-tracking/timeTrackingReducer.js [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+  hours: {},
+  displayHours: false
+};
+
+export default handleActions({
+  SET_DATES: (state, { dates }) => 
+    set('dates', dates, state )
+  ,
+  UPDATE_OFFSET: (state, { offset }) => {
+    if(offset === 7) {
```
Comment (ronlanger) [inline proj/src/reducers/time-tracking/timeTrackingReducer.js]: i would prefer having this logic in a utils function and probably in the component so the reducer has less logic in it

File: proj/src/store.js [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+const middleWare = [thunk]
+
+const initialState = {};
+
+const store = createStore(
+  rootReducer,
+  initialState,
+  compose(
+    applyMiddleware(...middleWare),
+    persistState(),
```
Comment (ronlanger) [inline proj/src/store.js:16]: does this work? you have your data saved when you refresh?

File: proj/src/views/Projects.js [hunk 11]
```diff
@@ -139,44 +79,43 @@ class Projects extends Component {
               <LayoutSelector onClick={() => this.changeLayout('grid')} src={gridLayout} alt="grid-icon"/>
               <LayoutSelector onClick={() => this.changeLayout('list')} src={listLayout} alt="list-icon"/>
             </LayoutSelectors>
-            <AddButton onClick={this.openEditModal}>ADD PROJECT</AddButton>
+            <AddButton onClick={() => this.props.openEditProject(null)}>ADD PROJECT</AddButton>
```
Comment (ronlanger) [inline proj/src/views/Projects.js]: why send null?

File: proj/src/views/TimeTracking.js [hunk 12]
```diff
@@ -17,22 +22,30 @@ class TimeTracking extends Component {
...
+    console.log(this.props)
+    if(prevProps.offset !== this.props.offset) {
+      this.renderDates()
+    }
+  }
+
+  handleChangeTime = (hoursValue, id) => {
+    // const hours = { ...this.state.hours };
+
+    // this.setState({
```
Comment (ronlanger) [inline proj/src/views/TimeTracking.js]: remove if not needed

File: proj/src/views/TimeTracking.js [hunk 13]
```diff
@@ -17,22 +22,30 @@ class TimeTracking extends Component {
     this.renderDates();
   }
 
-  changeTime = (hoursValue, id) => {
-    const hours = { ...this.state.hours };
+  componentDidUpdate(prevProps, prevState) {
+    console.log('prev', prevProps)
```
Comment (ronlanger) [inline proj/src/views/TimeTracking.js]: remove

**Review by ronlanger — CHANGES_REQUESTED** [review]

File: proj/src/store.js (hunk 10 above)
Comment (Ofir-Smuha) [reply proj/src/store.js:16]: Yep..

Comment (ronlanger) [reply proj/src/store.js:16]: 👍

File: proj/src/reducers/time-tracking/timeTrackingReducer.js [hunk 14]
```diff
@@ -1,6 +1,8 @@
-import { handleActions, combineActions } from 'redux-actions';
+import { handleActions } from 'redux-actions';
 import { set } from 'lodash/fp'
 
+import timeSet from './timeSetReducer'
```
Comment (ronlanger) [inline proj/src/reducers/time-tracking/timeTrackingReducer.js:4]: utils functions should sit under utils directory not in the reducer. the only case i can think of that this is ok is when the function sits in the reducer and serves that reducer alone

File: proj/src/views/Projects.js [hunk 15]
```diff
@@ -79,7 +79,7 @@ class Projects extends Component {
               <LayoutSelector onClick={() => this.changeLayout('grid')} src={gridLayout} alt="grid-icon"/>
               <LayoutSelector onClick={() => this.changeLayout('list')} src={listLayout} alt="list-icon"/>
             </LayoutSelectors>
-            <AddButton onClick={() => this.props.openEditProject(null)}>ADD PROJECT</AddButton>
+            <AddButton onClick={() => this.props.openEditProject()}>ADD PROJECT</AddButton>
```
Comment (ronlanger) [inline proj/src/views/Projects.js:82]: no need for an arrow function

File: proj/src/reducers/time-tracking/timeSetReducer.js [hunk 16]
```diff
@@ -0,0 +1,13 @@
+import { UPDATE_OFFSET } from 'actions/types'
```
Comment (ronlanger) [inline proj/src/reducers/time-tracking/timeSetReducer.js:1]: this should be in utils/timeTracking file or something of that sort

**Pull request merged**